### ISSUE TITLE
Consider switching to regex for content blocking

### ISSUE THREAD

**Issue by ahunt** [mozilla-mobile/focus-android#123]
I just learned iOS uses regex for content blocking (I'd assumed the browser did the same as the safari content blocking, but that's not the case). They use regex - I ended up building a trie-based system (which I think is what regex uses internally anyway - but regex is more complex):
https://github.com/mozilla-mobile/focus/blob/master/Blockzilla/LocalContentBlocker.swift#L174

We could consider switching to regex too, which would let us discard some code and also copy iOS's regex building (instead of our own blocklist parsing + trie construction). This might result in better performance, or might result in worse performance (our implementation supports a lot less than regex, so should be simpler and therefore faster - but Java regex could be implemented better / more efficient - I simply don't know).

This isn't needed for v1, and could possibly result in worse performance. It would be interesting to get a memory/performance comparison to determine that. Given that the current code works well (and has a full test suite), there's no real reason to change it - but it might save maintenance work in future (or it could be worse for maintenance, since it's regex).

(I'm filing this mainly for knowledge's sake, and I don't think there's any value in doing this anytime soon.)

**Comment by ahunt** [comment]
This isn't of huge importance, but I'm implementing an iOS-inspired UrlMatcher just to allow comparison of results - which should help catch bugs in the original version if present. I'm planning to run both matchers in parallel in debug (and probably beta) builds, and we can then throw exceptions (which results in crash reports) if results don't match.

**Comment by ahunt** [comment]
The good:
- My custom implementation is approx 100-200x faster than the iOS approach (I've ported the code as close to 1:1 as possible). I.e. up to 0.2ms per resource with the current code, compared to ~42ms per resource with the regex implementation (on a N6P).

The bad:
- There's one bug in the current implementation where we block too many sites:
-- twitter.com is defined as blocked in the blocklist (unless you're on a twitter property)
-- My intitial implementation also blocks ads-twitter.com (which is a  different domain), the iOS version (and ported code) doesn't. (I've got a fix for this.)

(ads-twitter.com should probably be on the blocklists, but that's a separate issue - and we should only block sites that are actually on the blocklist)

I've not had any other issues, I'm thinking it's probably best to enable both versions only in debug builds for now so we can see if any other bugs turn up - I'd still be keen to enable more widely on beta to see if there are any other bugs lurking in either version at some point.
